wgpu-core/src/conv.rs

```diff
@@ -446,28 +444,48 @@ pub fn map_texture_dimension_size(
         depth,
     }: wgt::Extent3d,
     sample_size: u32,
-) -> hal::image::Kind {
+) -> Result<hal::image::Kind, MapTextureDimensionSizeError> {
     use hal::image::Kind as H;
     use wgt::TextureDimension::*;
-    match dimension {
+    Ok(match dimension {
         D1 => {
-            assert_eq!(height, 1);
-            assert_eq!(sample_size, 1);
-            H::D1(width, checked_u32_as_u16(depth))
+            if height != 1 {
+                return Err(MapTextureDimensionSizeError::InvalidHeight);
+            }
+            if sample_size != 1 {
+                return Err(MapTextureDimensionSizeError::InvalidSampleCount(
+                    sample_size,
+                ));
+            }
+            H::D1(width, depth.try_into()?)
         }
         D2 => {
-            assert!(
-                sample_size <= 32 && is_power_of_two(sample_size),
-                "Invalid sample_count of {}",
-                sample_size
-            );
-            H::D2(width, height, checked_u32_as_u16(depth), sample_size as u8)
+            if sample_size > 32 || !is_power_of_two(sample_size) {
+                return Err(MapTextureDimensionSizeError::InvalidSampleCount(
+                    sample_size,
+                ));
+            }
+            H::D2(width, height, depth.try_into()?, sample_size as u8)
         }
         D3 => {
-            assert_eq!(sample_size, 1);
+            if sample_size != 1 {
+                return Err(MapTextureDimensionSizeError::InvalidSampleCount(
+                    sample_size,
+                ));
+            }
             H::D3(width, height, depth)
         }
-    }
+    })
+}
+
+#[derive(Clone, Debug, Error)]
+pub enum MapTextureDimensionSizeError {
+    #[error("1D images must have height set to 1")]
+    InvalidHeight,
+    #[error("sample count {0} is invalid")]
+    InvalidSampleCount(u32),
+    #[error(transparent)]
+    ValueTooLarge(#[from] TryFromIntError),
```

Comment:
This is meant to be for array sizes only, right?
We'll need to have that limit officially (on the array size), and this error would then be `TooManyArrayLayers(value)` or something like this. We shouldn't expose `TryFromIntError`, since the actual limit is always smaller than the u16::max

Reply:
the `TryFromIntError` is still here?

Reply:
Fixed now